src/test/ui/array-slice-vec/subslice-patterns-const-eval-match.rs

```diff
@@ -0,0 +1,110 @@
+// This test comprehensively checks the passing static and dynamic semantics
+// of subslice patterns `..`, `x @ ..`, `ref x @ ..`, and `ref mut @ ..`
+// in slice patterns `[$($pat), $(,)?]` .
+
+// run-pass
+
+#![feature(slice_patterns, const_fn, const_if_match)]
+#[derive(PartialEq, Debug, Clone)]
+struct N(u8);
+
+#[derive(PartialEq, Debug, Clone)]
+struct Z;
+
+macro_rules! n {
+    ($($e:expr),* $(,)?) => {
+        [$(N($e)),*]
+    }
+}
+
+macro_rules! zed {
+    ($e:expr) => { Z }
+}
+
+macro_rules! z {
+    ($($e:expr),* $(,)?) => {
+        [$(zed!($e)),*]
+    }
+}
+
+macro_rules! compare_evaluation {
+    ($e:expr, $t:ty $(,)?) => {{
+        const CONST_EVAL: $t = $e;
+        const fn const_eval() -> $t { $e }
+        const CONST_EVAL2: $t = const_eval();
+        assert_eq!(CONST_EVAL, $e);
```

Comment:
You should do `let x = $e;` and then use that value in `assert_eq`, because otherwise it may get promoted and const evaluated (not happening here due to the `if let`, but still.